core/domain/exp_jobs_one_off.py

```diff
@@ -528,6 +528,95 @@ def reduce(key, values):
         yield (key, values)
 
 
+class MathExplorationsImageGenerationAuditJob(
+        jobs.BaseMapReduceOneOffJobManager):
+    """Job that finds all the explorations with math rich text components and
+    creates a temporary storage model with all the information required for
+    generating math rich text component SVG images.
+    """
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [exp_models.ExplorationModel]
+
+    @staticmethod
+    def map(item):
+        if item.deleted:
+            return
+
+        exploration = exp_fetchers.get_exploration_from_model(item)
+        try:
+            exploration.validate()
+        except Exception as e:
+            logging.error(
+                'Exploration %s failed non-strict validation: %s' %
+                (item.id, e))
+            return
+        size_of_math_svg = 0
+        largest_math_expression = ''
+        list_of_latex_values = []
+        for state in exploration.states.values():
+            html_string = ''.join(state.get_all_html_content_strings())
+            (
+                size_of_math_svg_in_state, largest_math_expression_in_state,
+                list_of_latex_values_in_state) = (
+                    html_validation_service.
+                    extract_math_rich_text_related_information_from_html(
+                        html_string))
+            list_of_latex_values.extend(list_of_latex_values_in_state)
+            largest_math_expression = (
+                max(
+                    largest_math_expression,
+                    largest_math_expression_in_state, key=len)
+            )
+            size_of_math_svg = size_of_math_svg + size_of_math_svg_in_state
+
+        if size_of_math_svg > 0:
+            value_dict = {
+                'exploration_id': item.id,
+                'size_of_math_svgs': size_of_math_svg,
+                'largest_math_expression': largest_math_expression,
+                'latex_values': list(set(list_of_latex_values))
+            }
+            yield (
+                'Information about math-expressions in explorations',
+                value_dict)
+
+    @staticmethod
+    def reduce(key, values):
+        final_values = [ast.literal_eval(value) for value in values]
+        number_of_explorations_having_math = 0
+        largest_math_expression = ''
+        estimated_no_of_batches = 1
+        size_of_math_svgs_in_a_batch = 0
+        for value in final_values:
+            exp_models.ExplorationMathRichTextInfoModel(
```

Comment:
Woah, this doesn't scale at all. You should use put_multi().

Also this is just the audit job right? Audit jobs should not put anything into the datastore. Instead you should have some kind of regeneration job, similar to how we regenerate summaries etc. And maybe a deletion job to clear all existing ExplorationMathRichTextInfoModels so that you can do a clean start over (in case any data there becomes corrupted).

Reply:
done